src/coreclr/jit/lower.cpp

```diff
@@ -7752,6 +7777,66 @@ GenTree* Lowering::LowerSignedDivOrMod(GenTree* node)
     return node->gtNext;
 }
 
+//------------------------------------------------------------------------
+// TryFoldBinop: Try removing a binop node by constant folding.
+//
+// Parameters:
+//   node - the node
+//
+// Returns:
+//   True if the node was removed
+//
+bool Lowering::TryFoldBinop(GenTreeOp* node)
+{
+    if (node->gtSetFlags())
+    {
+        return false;
+    }
+
+    GenTree* op1 = node->gtGetOp1();
+    GenTree* op2 = node->gtGetOp2();
+
+    if (op1->IsIntegralConst() && op2->IsIntegralConst() &&
+        op1->AsIntConCommon()->ImmedValCanBeFolded(comp, node->gtOper) &&
+        op2->AsIntConCommon()->ImmedValCanBeFolded(comp, node->gtOper))
+    {
+        GenTree* folded = comp->gtFoldExprConst(node);
+        assert(folded == node);
+        if (!folded->OperIsConst())
+        {
+            return false;
+        }
+
+        op1->SetUnusedValue();
+        op2->SetUnusedValue();
+        return true;
+    }
+
+    if (node->OperIs(GT_LSH, GT_RSH, GT_RSZ, GT_ROL, GT_ROR, GT_OR, GT_XOR) &&
```

Comment:
sounds like what `gtFoldExprSpecial` does

Reply:
I presume the problem with calling `gtFold*` apis in Lower is that they must not create new nodes (other than what they return)

Reply:
Yes, I am trying to use the lowest level API I can for this to not run into issues calling this from lower... There's probably a better way to factor this folding to make that clear.

Reply:
Maybe it would be better to have something like
```cpp
bool gtFoldExprValue32(GenTree* node, int32_t* result);
bool gtFoldExprValue64(GenTree* node, int64_t* result);
```
and then call that from both `gtFoldExprConst` and lowering.

Reply:
I don't think I'm going to try changing this here. We should probably refactor `gtFold*` APIs to give it some idea of what kind of IR it is allowed to create instead, and then use it more directly from lowering.